app/pull-gitlab-variables.sh

```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+# arguments:
+# - ATLANTIS_GITLAB_TOKEN
+# - HEAD_REPO_OWNER
+# - HEAD_REPO_NAME
+# - ALLOWLIST_FILE e.g. GOOGLE_APPLICATION_CREDENTIALS
+# - ALLOWLIST_ENV_VAR e.g. AWS_ACCESS_KEY_ID,AWS_SECRET_ACCESS_KEY
+# - REPO_REL_DIR e.g environments/dev/cicd/azure
+
+ENV_SCOPE=$(echo $REPO_REL_DIR | sed -nE 's/.*\/(dev|test|prod|staging)\/.*/\1/p')
+${ENV_SCOPE:=*}
+
+
+ALLOWLIST_ENV_VAR_ARRAY=($(printf '%s\n' ${ALLOWLIST_ENV_VAR//,/ } | sort))
+ALLOWLIST_FILE_ARRAY=($(printf '%s\n' ${ALLOWLIST_FILE//,/ } | sort))
+
+# Variable needed for `glab`
+export GITLAB_TOKEN=${ATLANTIS_GITLAB_TOKEN}
+
+ENCODED_REPO_NAME=$(jq -rn --arg x "${HEAD_REPO_OWNER}/${HEAD_REPO_NAME}" '$x|@uri')
+API_RESPONSE=$(glab api projects/$ENCODED_REPO_NAME/variables)
+MULTIENV_RESULT=""
+
+
+ENV_VAR_VARIABLES=$(echo $API_RESPONSE | jq -c ".[] | select(.variable_type == \"env_var\" and (.environment_scope == \"$ENV_SCOPE\" or .environment_scope == \"*\")) | {key, value}" | jq -cs 'sort_by(.key) | .[]')
+for v in $ENV_VAR_VARIABLES
+do
+  name=$(echo $v | jq -r '.key')
+  content=$(echo $v | jq -r '.value')
+  MULTIENV_RESULT+="${name}=${content}",
+done
+
+ENV_VARS=$(echo $API_RESPONSE | jq -cr "[.[] | select(.variable_type == \"env_var\" and (.environment_scope == \"$ENV_SCOPE\" or .environment_scope == \"*\")).key] | sort | .[]")
+RESULT=($(comm -13 <(printf '%s\n' ${ALLOWLIST_ENV_VAR_ARRAY[@]}) <(printf '%s\n' ${ENV_VARS[@]})))
+
+if [ ${#RESULT[@]} -gt 0 ]; then
+  printf "Not allowed environment variables: [%s]\n" $(IFS=,; printf %s "${RESULT[*]}")
+  exit 1;
```

Comment:
I can see, that you're failing the whole script if "not allowed vars" are available - I'm not sure if that is an optimal approach.

I could imagine, that some other tool uses variable that we're banning... IMO we should remove blocked variable from the `ENV_VAR` array, output warning (with relevant information) but continue the script

Reply:
ok, fixed